File: apps/browser-demo/src/deviceAgent/provider/researchTargets.ts
export function researchTargetsForPayload(payload: Record<string, unknown>): ReadonlyArray<Record<string, unknown>> {
  const context = payload.context;
  if (!context || typeof context !== 'object' || Array.isArray(context)) return [];
  const targets = (context as Record<string, unknown>).researchTargets;
  if (!Array.isArray(targets)) return [];
  return targets.filter((target): target is Record<string, unknown> => (
    Boolean(target) && typeof target === 'object' && !Array.isArray(target)
  ));
}

File: apps/android-twa/app/src/main/java/com/agentic/wallet/agent/provider/ResearchTargets.kt
package com.agentic.wallet.agent.provider

import org.json.JSONArray
import org.json.JSONObject

internal fun researchTargetsForPayload(payload: JSONObject): JSONArray? {
    val context = payload.optJSONObject("context") ?: return null
    val targets = context.optJSONArray("researchTargets") ?: return null
    return if (targets.length() > 0) targets else null
}
